components/signoffs/signoffController.js

```diff
@@ -0,0 +1,171 @@
+    }
+  } catch (e) {
+    handleErrors(e, 'Error getting signoff(s) for case', logger, next);
+  }
+};
+
+const addSignoff = async (req, res, next) => {
+  try {
+    const existingSignoffs = await signoffDao.getByCaseIdentifier(
+      req.body.caseIdentifier
+    );
+
+    const validationResults = validateObjectsFromUser(req.body);
+    if (existingSignoffs == null || !existingSignoffs.length) {
+      const createdSignoff = await upsert(validationResults);
+      return res.status(201).json(createdSignoff);
+    } else {
+      for (const [, existingSignoff] of Object.entries(existingSignoffs)) {
+        if (
+          existingSignoff.signoffStepName ==
+            validationResults.signoffStepName &&
+          existingSignoff.deliverableType == validationResults.deliverableType
+        ) {
+          // signoff step is same, replace the old signoff record
+          const createdSignoff = await upsert(validationResults);
+          return res.status(200).json(createdSignoff);
+        } else {
+          // signoff data is different
+          const createdSignoff = await upsert(validationResults);
+          return res.status(201).json(createdSignoff);
+        }
+      }
+    }
+  } catch (e) {
+    handleErrors(e, 'Error adding sign-off', logger, next);
+  }
+};
+
+const upsert = (signoffInfo) => {
+  return signoffDao.addSignoff(signoffInfo);
+};
+
+//Currently are not exposing this as we do not want users
+//to delete sign-off records
+const deleteSignoff = async (req, res, next) => {
+  try {
+    if (!req.body.id)
+      throw Error(400, 'Error: no "signoffId" provided in request body');
+
+    const username = validateUsername(req.body.username);
+
+    const result = await signoffDao.deleteSignoff(req.body.id);
+    res.status(200).json(result);
+    next();
+  } catch (e) {
+    handleErrors(e, 'Error deleting sign-off records', logger, next);
+  }
+};
+
+function validateUsername (param) {
+  const user = nullifyIfBlank(param);
+  if (user == null || !user.length)
+    return new ValidationError('username must be provided');
+  if (user.match(/\W+/))
+    return new ValidationError('username must contain only letters');
+  return user;
+}
+
+function validateComment (param) {
+  let comment = nullifyIfBlank(param);
+  if (comment !== null)
+    comment = decodeURIComponent(comment.replace(/\+/g, ' '));
+  return comment;
+}
+
+function nullifyIfBlank (value) {
+  if (typeof value == 'undefined' || value === null || value.length == 0)
+    value = null;
+  return value;
+}
+
+function validateStepName (param) {
+  let stepname = nullifyIfBlank(param); //required by endpoint so shouldn't ever nullify
+  if (stepname !== 'undefined' && stepname !== null && stepname.length) {
+    stepname = stepname.toUpperCase();
+  }
+  let validSteps = ['ANALYSIS REVIEW', 'RELEASE APPROVAL', 'RELEASE'];
+  if (!validSteps.includes(stepname)) {
+    return new ValidationError(
+      'Sign-off must be associated with a valid step name: ' +
```

Comment:
Can you include `stepname` in the error message here and in the other error messages like this one? Something like `"signoff must be associated with a valid step name: " + validSteps.toString() + ", instead got " + stepname`

Reply:
Added, thanks